File: src/python/maps/map01.py
from browser import window
import ika

from yeti import Yeti


def AutoExec(engineRef):
    engineRef.mapThings.append(window.Snow.new(engineRef))
    if 'cowardrune' not in engineRef.saveFlags:
        engineRef.mapThings.append(RuneListener(engineRef))

def to2(engineRef):
    offset_from = 4 * 16  # first vertical pos possible
    offset_to = 38 * 16  # first vertical pos possible
    y = engineRef.player.y - offset_from + offset_to
    yield from engineRef.mapSwitchTask('map02.ika-map', (48 * 16, y))

def to49(engineRef):
    yield from engineRef.mapSwitchTask('map49.ika-map', (14 * 16, 23 * 16))

class RuneListener(object):
    def __init__(self, engineRef):
        self.engineRef = engineRef

    def update(self):
        if 'waterguard' in self.engineRef.saveFlags and 'fireguard' in self.engineRef.saveFlags and 'windguard' in self.engineRef.saveFlags:
            self.engineRef.addEntity(
                Yeti(self.engineRef, self.engineRef.map.addSprite(35 * 16, 19 * 16, self.engineRef.player.layer, 'yeti.ika-sprite')))
            return True

    def draw(self):
        pass

File: src/winter/maps/map01.ts
import { MapScript } from "./mapscript.js"
import { Thing } from "./../thing.js"
import { PyEngine, Snow } from "./../winter.js"
import { Yeti } from "./../yeti.js"

export default new MapScript(autoexec, {to2, to49})

class RuneListener extends Thing {
    constructor(private engineRef: PyEngine) { super() }

    update() {
        if (this.engineRef.getSaveFlag('waterguard') === '') { return false; }
        if (this.engineRef.getSaveFlag('fireguard') === '') { return false; }
        if (this.engineRef.getSaveFlag('windguard') === '') { return false; }

        const engine = this.engineRef.getEngine()
        const p = this.engineRef.getPlayerEntity()
        this.engineRef.pyAddEntity(
            new Yeti(this.engineRef, engine.map.addSprite(35 * 16, 19 * 16, p.sprite.layer, 'yeti.ika-sprite'))
        )
        return true
    }
}

function autoexec(engineRef: PyEngine) {
    engineRef.addMapThing(new Snow(engineRef))
    if (engineRef.getSaveFlag('cowardrune') === '') {
        engineRef.addMapThing(new RuneListener(engineRef))
    }
}

function *to2(engineRef: PyEngine) {
    const offset_from = 4 * 16  // first vertical pos possible
    const offset_to = 38 * 16  // first vertical pos possible
    const p = engineRef.getPlayerEntity()
    const y = p.sprite.y - offset_from + offset_to
    yield* engineRef.mapSwitchTask('map02.ika-map', [48 * 16, y])
}

function *to49(engineRef: PyEngine) {
    yield* engineRef.mapSwitchTask('map49.ika-map', [14 * 16, 23 * 16])
}
